### PR TITLE
Mini Scopes now allow movement with slowdown, can move scope direction with ctrl+WASD.

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
Allows marines to move while using scope zoom, but while doing so they get slowdown to compensate
and removes a SECONDS of scopes aiming movement slowdown(was multiplying it by 10, making scopes have 0.6 slowdown just by being on your gun)
Normal default scope does 1.5 slowdown(1.44 + 0.06 aim slowdown[wielded slowdown]) (disabled due to maptick)
and miniscope does 0.5 slowdown(0.46 + 0.04 aim slowdown)

## Why It's Good For The Game
Makes scopes be less clunky and allow people with them activated to move(slower), making you not need to reactivate scope every tile you move(which is horrible since changing viewsize on byond is laggy) and makes miniscopes more usefull.

## Changelog
:cl:
add: Added new things
balance: you can now move with mini scopes zoomed, which gives you a slowdown when zoomed to compensate.
balance: removes a multiplier that was making scopes give massive slowdown when just attached to gun.
/:cl:

### CODE REVIEW

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Carrying the scope slowdown removal OK
Moving scopes not OK due to maptick concerns as well as  buffing offscreening

**Comment by Chaznoodles** [comment]
Offscreen combat bad

**Comment by SplinterGP** [comment]
> Offscreen combat bad

You can alrready offscreen combat, you cant just move 1 tile as a sniper without committing seizure on youe screen

**Comment by SplinterGP** [comment]
> Carrying the scope slowdown removal OK
> Moving scopes not OK due to maptick concerns as well as buffing offscreening

>only miniscope allows movement.
Die

**Comment by Pariah919** [comment]
miniscope bad
this makes it okay
though honestly it should probably give you slowdown too

File: code/modules/projectiles/gun_attachables.dm [hunk 1]
```diff
@@ -739,17 +744,25 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
 		activate_attachment(user, TRUE)
 
 /obj/item/attachable/scope/onzoom(mob/living/user)
-	RegisterSignal(user, list(COMSIG_MOVABLE_MOVED, COMSIG_CARBON_SWAPPED_HANDS), .proc/zoom_item_turnoff)
+	if(!zoom_movement)
+		RegisterSignal(user, COMSIG_MOVABLE_MOVED, .proc/zoom_item_turnoff)
+	RegisterSignal(user, COMSIG_CARBON_SWAPPED_HANDS, .proc/zoom_item_turnoff)
 	RegisterSignal(master_gun, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_UNWIELD, COMSIG_ITEM_DROPPED), .proc/zoom_item_turnoff)
+	if(zoom_movement)
+		user.add_movespeed_modifier(MOVESPEED_ID_SCOPE_SLOWDOWN, TRUE, 0, NONE, TRUE, zoom_slowdown)
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_attachables.dm:752]: else exists

File: code/modules/projectiles/gun_attachables.dm [hunk 2]
```diff
@@ -739,17 +744,25 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
 		active_nightvision = TRUE
 
 /obj/item/attachable/scope/onunzoom(mob/living/user)
-	UnregisterSignal(user, list(COMSIG_MOVABLE_MOVED, COMSIG_CARBON_SWAPPED_HANDS))
+	if(!zoom_movement)
+		UnregisterSignal(user, COMSIG_MOVABLE_MOVED)
+	UnregisterSignal(user, COMSIG_CARBON_SWAPPED_HANDS)
 	UnregisterSignal(master_gun, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_UNWIELD, COMSIG_ITEM_DROPPED))
+	if(zoom_movement)
+		user.remove_movespeed_modifier(MOVESPEED_ID_SCOPE_SLOWDOWN)
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_attachables.dm:765]: else exists

**Review by TiviPlus — CHANGES_REQUESTED** [review]
image

File: code/modules/projectiles/gun_attachables.dm [hunk 3]
```diff
@@ -689,7 +690,10 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
 	var/zoom_offset = 11
 	///how many tiles to increase the users view box
 	var/zoom_viewsize = 12
-	scoped_accuracy_mod = SCOPE_RAIL
+	///can this scope move while zooming ? default of no, scopes with higher viewrange than 7 cause maptick lag on moving according to tivi.
+	var/zoom_movement = FALSE
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_attachables.dm:694]: also add a CRASH()  for scopes if they set it too high

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_attachables.dm (hunk 3 above)
Comment (SplinterGP) [reply code/modules/projectiles/gun_attachables.dm:694]: if zoom movement with viewrange high ?

Comment (TiviPlus) [reply code/modules/projectiles/gun_attachables.dm:694]: if they have both this and a viewrange higher than default add a CRASH to the zoom proc

**Comment by TiviPlus** [comment]
Still pending

**Comment by TiviPlus** [comment]
Pending fix of binocular zooming bug

File: code/game/objects/items.dm [hunk 4]
```diff
@@ -749,17 +735,71 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 	else if(!istype(src, /obj/item/attachable/scope))
 		user.set_interaction(src)
 
+/obj/item/proc/canzoom(mob/living/user)
+	var/zoom_device = zoomdevicename ? "\improper [zoomdevicename] of [src]" : "\improper [src]"
```
Comment (TiviPlus) [inline code/game/objects/items.dm:739]: Autodoc

File: code/modules/mob/mob.dm [hunk 5]
```diff
@@ -584,11 +584,20 @@
...
 		buckled.setDir(ndir)
 	return TRUE
 
+/mob/setDir(newdir)
+	if(facing_dir)
+		if(!canface())
+			facing_dir = null
+		else if(newdir != facing_dir)
+			return
+	..()
```
Comment (TiviPlus) [inline code/modules/mob/mob.dm:600]: what the fuck is thiiiiis

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/mob.dm (hunk 5 above)
Comment (SplinterGP) [reply code/modules/mob/mob.dm:600]: stops user from changing their direction when facing_dir is set, used by scopes

Comment (TiviPlus) [reply code/modules/mob/mob.dm:600]: As discussed in discord, change this

**Comment by MiniMeatwad** [comment]
Should remove the ability to move while scoped with the SADAR

**Pull request closed without merging**